src/librustc_trans/trans/meth.rs

```diff
@@ -866,3 +867,20 @@ pub fn trans_trait_cast<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
 
     bcx
 }
+
+/// Replace the self type (&Self or Box<Self>) with an opaque pointer.
+pub fn opaque_method_ty<'tcx>(tcx: &ty::ctxt<'tcx>, method_ty: &ty::BareFnTy<'tcx>)
+        -> ty::BareFnTy<'tcx> {
```

Comment:
Would be nice if this interned the `BareFnTy` itself - though I'm not sure we've agreed on a style, I believe we are heading towards passing only `&'tcx Foo<'tcx>` around (which eventually turns into a `Foo<'tcx>` alias for the reference type - just like `Ty<'tcx>`).